File: ruby/spec/data-structures/linked-list/singly_linked_list_spec.rb
require_relative '../../spec_helper'
describe SinglyLinkedList do
  let(:head) { "head" }
  let(:node) { "next" }
  let(:list) { SinglyLinkedList.new head}
  context "#head" do
    it "defaults to nil" do
      list = SinglyLinkedList.new
      expect(list.head).to be_nil
    end
    it "can be set with initial data" do
      expect(list.head.data).to eq head
    end
  end
  context "#tail" do
    it "returns the element at the end of the list" do
      expect(list.tail.data).to eq head
    end
  end
  context "#push_bottom" do
    it "sets head if it's nil" do
      list = SinglyLinkedList.new
      list.push_bottom(head)
      expect(list.head.data).to eq head
    end
    it "head is not affected if it's already set" do
      list.push_bottom(node)
      expect(list.head.data).to eq head
    end
    it "push_bottoms the node to the end of the list" do
      list.push_bottom(node)
      tail = "tail"
      list.push_bottom(tail)
      expect(list.head.next.next.data).to eq tail
    end
    it "is the tail element" do
      list.push_bottom(node)
      tail = "tail"
      list.push_bottom(tail)
      expect(list.tail.data).to eq tail
    end
  end

  context "#push_top" do
    it "sets head if it's nil" do
      list = SinglyLinkedList.new
      list.push_top(head)
      expect(list.head.data).to eq head
    end
    it "sets the head if it's already set" do
      list.push_top(node)
      expect(list.head.data).to eq node
    end
    it "head is the tails element" do
      list.push_top(node)
      expect(list.head.data).to eq node
      expect(list.tail.data).to eq head
    end
  end
  context "#find" do
    it "returns the node if data is found" do
      expect(list.find(head).data).to eq head
      list.push_bottom(node)
      expect(list.find(node).data).to eq node
    end
    it "returns nil if data not found" do
      expect(list.find("not there")).to be_nil
    end
  end

  context "#remove" do
    it "removes the head if its the give data" do
      list.push_bottom(node)
      list.remove head
      expect(list.find(head)).to be_nil
      expect(list.head.data).to eq node
    end
    it "removes the node if data is not at head" do
      list.push_bottom("another")
      list.push_bottom(node)
      list.remove node
      expect(list.find(node)).to be_nil
      expect(list.find("another").data).to eq "another"
      expect(list.find(head).data).to eq head
    end
  end
  context "#map" do
    it "yields the block to each element" do
      expect(list.map {|node| node.data }).to eq ["head"]
    end
  end
end

File: javascript/test/data-structures/linked-list/singlyLinkedListSpec.js
require('../../test_helper');
describe('SinglyLinkedList', function(){
  var data = "sweet data",
      list,
      node;
  beforeEach(function () {
    list = new SinglyLinkedList;
    node = new Node(data)
  });
  describe('#head', function(){
    it('is undefined on initialize', function(){
      expect(list.head).to.be.an('undefined');
    });
  });
  describe('#addBottom', function () {
    it('sets head if head is undefined', function () {
      list.addBottom(node)
      expect(list.head).to.eq(node);
    });
    it('adds to the bottom', function () {
      list.addBottom(node);
      var anotherNode = new Node("another node");
      list.addBottom(anotherNode);
      expect(list.head).to.eq(node);
      expect(list.head.next).to.eq(anotherNode);
      expect(list.find("another node")).to.eq(anotherNode);
    });
  });

  describe('#find', function () {
    it('returns undefined if no element found', function () {
      expect(list.find("yo")).to.be.an('undefined');
    });
  });

  describe('#addTop', function () {
    it('sets head', function () {
      list.addTop(node)
      expect(list.head).to.eq(node);
    });
    it('shifts the list', function () {
      list.addTop(node);
      var anotherNode = new Node("another node");
      list.addTop(anotherNode);
      expect(list.head).to.eq(anotherNode);
      expect(list.head.next).to.eq(node);
    });
  });

  describe('#remove', function () {
    it('removes head', function () {
      list.addBottom(node);
      list.remove(node.data);
      expect(list.head).to.be.an('undefined');
    });
  });
  it('removes the element if not head', function () {
    list.addBottom(node);
    var anotherNode = new Node("another node");
    list.addBottom(anotherNode);
    list.remove(anotherNode.data);
    expect(list.head).to.be.eq(node);
    expect(list.head.next).to.be.an('undefined');
    expect(list.find(anotherNode.data)).to.be.an('undefined');
    expect(list.find(node.data)).to.be.eq(node);
  });
})
